test/Tester/MemoryStorageProviderTests.cs

```diff
@@ -41,5 +42,79 @@ public class MemoryStorageProviderTests : HostedTestClusterEnsureDefaultStarted
             Assert.Equal("Alice", names[1]);
         }
+
+        [Fact, TestCategory("BVT"), TestCategory("Functional"), TestCategory("Storage")]
+        public async Task MemoryStorageGrainEnforcesEtagsTest()
+        {
+            var memoryStorageGrain = GrainClient.GrainFactory.GetGrain<IMemoryStorageGrain>(random.Next());
+
+            // Delete grain state from empty grain, should be safe.
+            await memoryStorageGrain.DeleteStateAsync("stateStore", "grainStoreKey", "eTag");
+
+            // Read grain state from empty grain, should be safe, but return nothing.
+            IGrainState grainState = await memoryStorageGrain.ReadStateAsync("stateStore", "grainStoreKey");
+            Assert.Null(grainState);
+
+            // write state with etag, when there is nothing in storage.  Most storage should fail this, but memory storage should succeed.
+            await memoryStorageGrain.WriteStateAsync("grainType", "grainId", TestGrainState.CreateRandom());
+
+            // write new state with null etag
+            string newEtag = await memoryStorageGrain.WriteStateAsync("grain", "id", TestGrainState.CreateWithEtag(null));
+            Assert.NotNull(newEtag);
+
+            // try to write new state with null etag;
+            await Assert.ThrowsAsync<InconsistentStateException>(() => memoryStorageGrain.WriteStateAsync("grain", "id", TestGrainState.CreateWithEtag(null)));
+
+            // try to write new state with different etag;
+            await Assert.ThrowsAsync<InconsistentStateException>(() => memoryStorageGrain.WriteStateAsync("grain", "id", TestGrainState.CreateWithEtag(newEtag+"a")));
+
+            // Write new state with good etag;
+            string latestEtag = await memoryStorageGrain.WriteStateAsync("grain", "id", TestGrainState.CreateWithEtag(newEtag));
+            Assert.NotNull(latestEtag);
+
+            // try delete state with null etag
+            await Assert.ThrowsAsync<InconsistentStateException>(() => memoryStorageGrain.DeleteStateAsync("grain", "id", null));
+
+            // try delete state with wrong etag
+            await Assert.ThrowsAsync<InconsistentStateException>(() => memoryStorageGrain.DeleteStateAsync("grain", "id", latestEtag+"a"));
+
+            // delete state
+            await memoryStorageGrain.DeleteStateAsync("grain", "id", latestEtag);
+
+            // Read grain deleted grain state, should be safe, but return nothing.
+            grainState = await memoryStorageGrain.ReadStateAsync("grain", "id");
+            Assert.Null(grainState);
+
+            // try delete already deleted grain state
+            await Assert.ThrowsAsync<InconsistentStateException>(() => memoryStorageGrain.DeleteStateAsync("grain", "id", latestEtag));
+
+            // try to write state to deleted state.
+            await Assert.ThrowsAsync<InconsistentStateException>(() => memoryStorageGrain.WriteStateAsync("grain", "id", TestGrainState.CreateWithEtag(latestEtag)));
+        }
```

Comment:
lastly, after delete, assert that you can now write with a null eTag (since it's the scenario we talked about in the previous comment)